src/core/python2js.c

```diff
@@ -372,58 +445,65 @@ _python2js_cache(PyObject* x, PyObject* map, int depth)
   } else {
     result = _python2js(x, map, depth);
   }
-  Py_DECREF(id);
   return result;
+finally:
+  return NULL;
 }
 
 JsRef
 python2js(PyObject* x)
 {
   RETURN_IF_SUCCEEDS(_python2js_immutable(x));
+  FAIL_IF_ERR_OCCURRED();
+  RETURN_IF_SUCCEEDS(_python2js_proxy(x));
+  FAIL_IF_ERR_OCCURRED();
   RETURN_IF_SUCCEEDS(pyproxy_new(x));
-  pythonexc2js();
+finally:
+  if (PyErr_Occurred()) {
+    if (!PyErr_ExceptionMatches(conversion_error)) {
+      _PyErr_FormatFromCause(conversion_error,
+                             "Conversion from python to javascript failed");
+    }
+  } else {
+    PyErr_SetString(internal_error,
+                    "Internal error occurred in python2js_with_depth");
+  }
+  return NULL;
 }
 
 JsRef
 python2js_with_depth(PyObject* x, int depth)
```

Comment:
Could we add some docstrings for these functions while we are at it?

Reply:
Okay, I added some docstrings and tried to improve names of things a bit.